### PR TITLE
pyterm: add native support

### PR DESCRIPTION
**Pull request by OlegHahm**
### Contribution description

I always found it annoying that native is lacking the convenience features of `pyterm` (e.g., history, concatenating of commands, logging etc). Of course, one could pipe stdin/stdout into something like socat and use pyterm's feature to connect to a TCP server (e.g., `socat EXEC:${NATIVEBIN},end-close,stderr,pty TCP-LISTEN:4711,reuseaddr,fork & pyterm -ts 4711`) but that's somewhat tedious and not well-aligned with the typical usage.

Hence, this PR will modify the `term` make target for native so that pyterm is called instead of calling the native binary directly. In turn, pyterm now has a new option which will spawn RIOT native processes as a subprocess (i.e., child process) and communicate to it via pipes.

### Testing procedure

Call `BOARD=native make term` for arbitrary RIOT applications.

Caution: this is not well tested so far - expect the unexpected.

### CODE REVIEW

**Comment by miri64** [comment]
Setting the `CI: run tests` label to make sure they are run on CI execution (I can imagine this PR will make some trouble there...). Other than that it looks fine from my side.

**Comment by benpicco** [comment]
> Setting the `CI: run tests` label to make sure they are run on CI execution (I can imagine this PR will make some trouble there...). Other than that it looks fine from my side.

IIRC that is only for hardware based tests, all `native` tests should always be run as part of CI

**Comment by OlegHahm** [comment]
Okay to squash and remove the WIP label?

**Comment by miri64** [comment]
Yepp, Murdock will take a while anyways, but `tools-test` now passes.

**Comment by OlegHahm** [comment]
I realized that this change breaks when passing arguments to native (like when using `USE_ZEP=1`). :-(

**Comment by miri64** [comment]
Mh... maybe, instead of passing parameter by parameter just have a “native parameters” parameter, to which you pass all parameters required for native. This way you can keep the old `TERMFLAGS` and just put them into the new “native parameters” parameter.

**Comment by miri64** [comment]
Of course that makes `port` somewhat unnecessary... not sure this is easily made optional

**Comment by OlegHahm** [comment]
```Make
ifndef TERMPROG
  TERMFLAGS += -n $(FLASHFILE)
endif
ifneq (,$(findstring pyterm,$(TERMPROG)))
  TERMFLAGS += -n $(FLASHFILE)
endif
```
I couldn't find any non-redundant version but would be happy if anyone had an idea.

File: dist/tools/pyterm/pyterm [hunk 1]
```diff
@@ -891,6 +931,8 @@ if __name__ == "__main__":
                         action="store_false",
                         help="Do not try to reconnect when failing on "
                              "connection setup (Default)")
+    parser.add_argument("native_tap", nargs=1)
```
Comment (miri64) [inline dist/tools/pyterm/pyterm:934]: IIRC, there are and may be applications that use more than one tap interface.

Comment (benpicco) [reply dist/tools/pyterm/pyterm:934]: e.g. `examples/gnrc_networking_subnets`

Comment (OlegHahm) [reply dist/tools/pyterm/pyterm:934]: Thanks for the hint. I will look into that.

File: boards/native/Makefile.include [hunk 2]
```diff
@@ -14,7 +14,23 @@ RESET ?= $(RIOTBOARD)/native/dist/reset.sh
...
+endif
+ifneq (,$(findstring pyterm,$(TERMPROG)))
+  SETPYTERMFLAGS = 1
+endif
+
+ifeq (1, $(SETPYTERMFLAGS))
+  TERMFLAGS += -n $(FLASHFILE)
+else
+  TERMPROG ?= $(FLASHFILE)
+endif
```
Comment (OlegHahm) [inline boards/native/Makefile.include:33]: That was my effort to keep application Makefiles as much as possible as they are - but that's overly complicated. I'm overhauling this.

File: examples/gnrc_networking/Makefile [hunk 3]
```diff
@@ -47,11 +47,11 @@ USE_ZEP ?= 0
...
 ifeq (1,$(USE_ZEP))
-  TERMFLAGS += -z [::1]:$(ZEP_PORT_BASE)
+  TERMFLAGS += --native-args '-z [::1]:$(ZEP_PORT_BASE)'
   USEMODULE += socket_zep
 
   ifneq (,$(ZEP_MAC))
-    TERMFLAGS += --eui64=$(ZEP_MAC)
+    TERMFLAGS += --native-args '\-\-eui64=$(ZEP_MAC)'
   endif
 endif
```
Comment (benpicco) [inline examples/gnrc_networking/Makefile:56]: If you are touching all the application makefiles anyway, you could also move this to a common place :innocent:

Comment (OlegHahm) [reply examples/gnrc_networking/Makefile:56]: I would be happy to comply but the actual added flags seem to be pretty diverse.

Comment (OlegHahm) [reply examples/gnrc_networking/Makefile]: I came up with a proposal for that. I also reverted to call native's binary directly for tests since pyterm isn't beneficial here and just adds overhead.

**Comment by miri64** [comment]
I was just thinking: Is this approach really just `native` specific? Isn't it just “wrap around an executable with arguments”? If yes, we could e.g. also wrap `ethos` or `sliptty`. All I'd ask for that is, that we remove the name "`native`" from pyterm and use "executable" or something instead. The porting to `ethos` or `sliptty` can be done later.

**Comment by OlegHahm** [comment]
May I rebase on master after #20178 has been merged?

File: examples/gnrc_border_router/Makefile.native.conf [hunk 4]
```diff
@@ -34,15 +34,11 @@ endif
...
+RIOT_TERMINAL ?= native
 TERMPROG ?= sudo $(RIOTTOOLS)/zep_dispatch/start_network.sh $(TERMPROG_FLAGS)
 
 # -z [::1]:$PORT for each ZEP device
 TERMFLAGS ?= $(patsubst %,-z [::1]:%, $(shell seq $(ZEP_PORT_BASE) $(ZEP_PORT_MAX)))
 
-# set optional ZEP l2 address
-ifneq (,$(ZEP_MAC))
-  TERMFLAGS += --eui64=$(ZEP_MAC)
-endif
```
Comment (miri64) [inline examples/gnrc_border_router/Makefile.native.conf:45]: Huh? Is this removed intentionally?

File: examples/gnrc_networking/Makefile [hunk 5]
```diff
@@ -47,12 +47,7 @@ USE_ZEP ?= 0
 # set the ZEP port for native
 ZEP_PORT_BASE ?= 17754
 ifeq (1,$(USE_ZEP))
-  TERMFLAGS += -z [::1]:$(ZEP_PORT_BASE)
   USEMODULE += socket_zep
-
-  ifneq (,$(ZEP_MAC))
-    TERMFLAGS += --eui64=$(ZEP_MAC)
-  endif
```
Comment (miri64) [inline examples/gnrc_networking/Makefile:55]: This as well?

File: tests/core/sched_change_priority/Makefile [hunk 6]
```diff
@@ -4,9 +4,6 @@ USEMODULE += nice
 USEMODULE += ps
 USEMODULE += shell_cmds_default
 
-# Use a terminal that does not introduce extra characters into the stream.
-RIOT_TERMINAL ?= socat
```
Comment (miri64) [inline tests/core/sched_change_priority/Makefile:8]: And this?

File: tests/sys/shell_coreclk/Makefile [hunk 7]
```diff
@@ -3,10 +3,4 @@ include ../Makefile.sys_common
 USEMODULE += shell
 USEMODULE += shell_cmd_coreclk
 
-# Use a terminal that does not introduce extra characters into the stream.
-RIOT_TERMINAL ?= socat
-
 include $(RIOTBASE)/Makefile.include
-
-# the test script skips tests if socat is not used
-$(call target-export-variables,$(RIOT_TERMINAL),RIOT_TERMINAL)
```
Comment (miri64) [inline tests/sys/shell_coreclk/Makefile:12]: And this?

**Comment by OlegHahm** [comment]
> (and also you can still follow the history with the force-push notifications nowadays)

I learned (or at least that is what I understood) that this only works if I just rebase within my own branch.

File: tests/net/socket_zep/tests/01-run.py [hunk 8]
```diff
@@ -51,9 +51,10 @@ def testfunc(child):
 
 
 if __name__ == "__main__":
-    os.environ['TERMFLAGS'] = "--eui64=00:5a:45:50:0a:00:30:38 -z [%s]:%d,[%s]:%d" % (
-            zep_params['local_addr'], zep_params['local_port'],
-            zep_params['remote_addr'], zep_params['remote_port'])
+    os.environ['TERMFLAGS'] = "--eui64=00:5a:45:50:0a:00:30:38 -z [%s]:%d,[%s]:%d" % (zep_params['local_addr'],
+                                                                                      zep_params['local_port'],
+                                                                                      zep_params['remote_addr'],
+                                                                                      zep_params['remote_port'])
```
Comment (miri64) [inline tests/net/socket_zep/tests/01-run.py:57]: Can you flake8 this please? I have the feeling this will be too long (+ I the change seems unrelated).

Comment (OlegHahm) [reply tests/net/socket_zep/tests/01-run.py:57]: You're right - it's unrelated (or rather a result of a back-and-forth). Flake seems to be cool with it, but I will revert it anyway.

File: tests/net/emcute/Makefile [hunk 9]
```diff
@@ -6,7 +6,7 @@ export TAP ?= tap0
 
 # use Ethernet as link-layer protocol
 ifeq (native,$(BOARD))
-  TERMFLAGS ?= $(TAP)
+  PORT += $(TAP)
```
Comment (benpicco) [inline tests/net/emcute/Makefile:9]: Shouldn't this be 
```suggestion
  PORT ?= $(TAP)
```

File: dist/tools/pyterm/pyterm [hunk 10]
```diff
@@ -891,6 +931,8 @@ if __name__ == "__main__":
                         action="store_false",
                         help="Do not try to reconnect when failing on "
                              "connection setup (Default)")
+    parser.add_argument("process_tap", nargs='*')
```
Comment (miri64) [inline dist/tools/pyterm/pyterm:934]: Does this still make sense? Why isn't this passed together with `-pa`?

Comment (OlegHahm) [reply dist/tools/pyterm/pyterm:934]: Hm, good question. I thought native was expecting the tap parameter as last parameter but apparently it does also work in between. So, yes, could be probably removed.

Comment (benpicco) [reply dist/tools/pyterm/pyterm:934]: I think it will then ignore the parameters that come after

Comment (OlegHahm) [reply dist/tools/pyterm/pyterm]: Nope, all are handled, e.g., using `gnrc_networking` with `USE_ZEP=1` and `USEMODULES += netdev_tap`
```
bin/native/gnrc_networking.elf  -z '[::1]:17754' tap0 -UAB
RIOT native interrupts/signals initialized.
RIOT native board initialized.
RIOT native hardware initialization complete.
...
ifconfig
...
Iface  8  HWaddr: 00:00  Channel: 26  NID: 0x23  PHY: O-QPSK 
          Long HWaddr: AB:00:00:00:00:00:00:00 
```
I.e., -z and -U are handled correctly.

Comment (miri64) [reply dist/tools/pyterm/pyterm]: And even if: Since the make system provides the parameters in most cases, any constraints on them can be ensured there.

**Comment by OlegHahm** [comment]
Anything left to be addressed?

**Comment by miri64** [comment]
What about https://github.com/RIOT-OS/RIOT/pull/20172#discussion_r1430091811 and https://github.com/RIOT-OS/RIOT/pull/20172/files#diff-7cd56ff235fec91ccddcc782a87833f544e77feb13e0e9a2d8c3043018439495?

**Comment by OlegHahm** [comment]
Seemed to be unnecessary and would probably complicate the configuration.

**Comment by miri64** [comment]
> Seemed to be unnecessary and would probably complicate the configuration.

Mh, but the comment says why socat was used: To not have extra characters in the stream. I can imagine that the tests depend on that. If you insist on `pyterm` at least set the `pyterm` flags that are used with the `cleanterm` target.

**Comment by miri64** [comment]
Just for that I set `run tests` to also test if the behavior non-`native` is preserved ;-).

**Comment by OlegHahm** [comment]
> Mh, but the comment says why socat was used: To not have extra characters in the stream. 

I'm under the impression that this comment is obsolete.

> I can imagine that the tests depend on that. If you insist on `pyterm` at least set the `pyterm` flags that are used with the `cleanterm` target.

But as far as I understand the tests are executed using the `cleanterm` target anyway.

The tests run successful on my PC and in the CI.

From the test code I also see no reason why the tests should fail with pyterm.

**Comment by miri64** [comment]
Since Murdock does not seem to have a problem with that either, I am fine with that explanation.

**Review by benpicco — APPROVED** [review]
This makes `native` so much nicer :smiley: 

I also don't see how the tests would depend on `pyterm` - they use their own python based testing harness (e.g. RIOTCtrl).

**Comment by benpicco** [comment]
Uh looks like `tests/build_system/test_tools` did not like this

**Comment by benpicco** [comment]
> I also don't see how the tests would depend on `pyterm`

Turns out I was wrong - or rather they depend on the *absence* of `pyterm`. (As in they don't use `pyterm` on the target boards either, but now will always use it for `native`)

Now for some reason CI did *not* catch this, but an [unrelated PR](https://github.com/RIOT-OS/RIOT/pull/20211) manages to [shake them out](https://ci.riot-os.org/details/9b9947d6b7b14c7b829272316f527555).

**Comment by benpicco** [comment]
arg this also broke `make debug` on `native`
